src/devices/LedDisplay/Ht16k33.cs

```diff
@@ -0,0 +1,230 @@
+    /// HT16K33 LED display driver
+    /// </summary>
+    /// <remarks>
+    /// Datasheet: https://www.adafruit.com/datasheets/ht16K33v110.pdf
+    /// Sources:
+    /// https://github.com/adafruit/Adafruit_LED_Backpack/blob/master/Adafruit_LEDBackpack.cpp
+    /// https://github.com/sobek1985/Adafruit_LEDBackpack/blob/master/Adafruit_LEDBackpack/AlphaNumericFourCharacters.cs
+    /// </remarks>
+    public abstract class Ht16k33 : IDisposable
+    {
+        #region Constants
+        /// <summary>
+        /// HT16K33 default I2C address
+        /// </summary>
+        public const int DefaultI2cAddress = 0x70;
+        #endregion
+
+        #region Protected members
+        /// <summary>
+        /// I2C device interface
+        /// </summary>
+        protected I2cDevice I2cDevice;
+        #endregion
+
+        #region Private members
+        #region Enums
+        private enum Command : byte
+        {
+            /// <summary>
+            /// HT16K33 blink command
+            /// </summary>
+            Blink = 0b1000_0000,
+            /// <summary>
+            /// HT16K33 blink display on command
+            /// </summary>
+            BlinkDisplayOn = 0b0000_0001,
+            /// <summary>
+            /// HT16K33 turn on oscillator command
+            /// </summary>
+            TurnOnOscillator = 0b0010_0001,
+            /// <summary>
+            /// HT16K33 set brightness command
+            /// </summary>
+            SetBrightness = 0b1110_0000,
+            /// <summary>
+            /// HT16K33 brightness bit mask
+            /// </summary>
+            BrightnessMask = 0b0000_1111
+        }
+        #endregion
+
+        /// <summary>
+        /// Display on/off state
+        /// </summary>
+        private bool _displayOn = true;
+
+        /// <summary>
+        /// Initial brightness level
+        /// </summary>
+        private byte _brightness;
+
+        /// <summary>
+        /// Initial blinking rate
+        /// </summary>
+        private BlinkRates _blinkRate;
+
+        /// <summary>
+        /// Send turn on oscillator command to device
+        /// </summary>
+        private void TurnOnOscillator() => I2cDevice.WriteByte((byte)Command.TurnOnOscillator);
+
+        /// <summary>
+        /// Send set brightness command to device
+        /// </summary>
+        private void SetBrightness() => I2cDevice.WriteByte((byte)((byte)Command.SetBrightness | (_brightness & (byte)Command.BrightnessMask)));
+
+        /// <summary>
+        /// Send Blink/DisplayOn command to device
+        /// </summary>
+        private void SetBlinkDisplayOn() => I2cDevice.WriteByte((byte)((byte)Command.Blink | (byte)Command.BlinkDisplayOn | ((byte)_blinkRate << 1)));
+        #endregion
+
+        #region Public members
+        #region Enums
+        /// <summary>
+        /// Blink rates
+        /// </summary>
+        /// <remarks>Specific to LED driver</remarks>
+        public enum BlinkRates : byte
```

Comment:
this should be private or not nested enum

Reply:
reason for not nested is that it's kinda annoying to write a type name for nested types since you always have to write things like:  `Htk16k33.BlinkRates.Off` instead of just `BlinkRates.Off`

Reply:
I know it's annoying, but it's specific to the driver, so I thought it should be namespaced somehow with the driver. I'll try to add some abstraction to this to resolve.

Reply:
Other option is `Htk16k33BlinkRate` (btw this should be singular unless it's `[Flags]`)